src/client.c

```diff
@@ -35,6 +35,56 @@ static const char *hdr2hex(const struct MBuf *data, char *buf, unsigned buflen)
 	return bin2hex(bin, dlen, buf, buflen);
 }
 
+/*
+ * An helper function to select authentication database for the current
+ * client. The order of preference will be as follows:
+ *   1) client->db->auth_dbname: Client's target database may alraedy have
+ * 		an auth_database defined. Prefer this when available.
+ *   2) cf_auth_dbname: Global authentication database name.
+ *   3) client->db: Client's target database. 
+ * NOTE: If authentication database is found but it is disabled, then
+ * 		 client will get disconnected. 
+ * Returns:
+ *   auth_database, based on the above preference. May return NULL when
+ *   				auth_db is disabled.
+ */
```

Comment:
> The chosen variable names are very similar (auth_db and auth_dbname)
Any suggestions for better ones? To me they seem quite clear, one is the db object, the other is a name. They are also inline with other functions that we already have like use_client_socket, where we have `db` and `dbname`.

> There are several returns.

Why is this an issue? I suggested the early-return pattern because I thought the initial version was quite hard to follow.

> I added a comment on the other PR https://github.com/pgbouncer/pgbouncer/pull/645 that I'm concerned about a fallback option. IMO it should be predictable. If the authentication database could not be found, it should fail.

I don't have a very strong opinion either way. But failing hard in case of wrong config, indeed has my slight preference.

Reply:
The early-return pattern is not an issue per se. I just said this in hopes that someone else could come up with something easier to read. We can always make hard-to-read code easier to read by adding comments.

Reply:
To confirm, is the ask here to make `auth_dbname` a breaking change between versions that when not set (in connection string or globally), it will cause disconnection?

The current behavior in that case is to use `dbname` in the connection string.

Reply:
No, (afaik) the ask is: if it is set, but the database is not defined, it should error. Right now it it falls back to use the `dbname` in that case.

So keep this:
```c
	/* use the client's target database if auth_dbname is not configured. */
	if (!auth_dbname) {
		return auth_db;
	}
```
but remove this
```c
	/* if auth_dbname is configured, attemp to find it.
	 * when database is found but it is disabled, it
	 * will result in client disconnection.
	 */
	auth_db = find_database(auth_dbname);
	if (!auth_db) {
		slog_info(client, "Authentication database %s could not be found. Reverting to client database.", auth_dbname);
		return client->db;
	}
```

Reply:
removed the logic of falling back. Instead, it will disconnect the client if `auth_db` cannot be found.